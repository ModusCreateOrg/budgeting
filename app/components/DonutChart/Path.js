import React, { Component, PropTypes } from 'react';

import { select, interpolate } from 'd3';

class Path extends Component {
  static propTypes = {
    data: PropTypes.object.isRequired,
    fill: PropTypes.string.isRequired,
    arcFn: PropTypes.func.isRequired,
    animDuration: PropTypes.number,
  };

  static defaultProps = {
    animDuration: 1000,
  };

  componentDidMount() {
    const { data, arcFn, animDuration } = this.props;
    const path = select(this.pathRef);
    const interpolateArc = interpolate(
      { startAngle: 0, endAngle: 0 },
      { startAngle: data.startAngle, endAngle: data.endAngle }
    );

    path.transition().duration(animDuration).attrTween('d', () => t => arcFn(interpolateArc(t)));
  }

  handleRefUpdate = ref => {
    this.pathRef = ref;
  };

  render() {
    const { data, arcFn, fill } = this.props;

    return <path ref={this.handleRefUpdate} fill={fill} d={arcFn(data)} />;
  }
}

export default Path;
