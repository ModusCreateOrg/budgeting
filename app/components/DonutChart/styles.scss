.donutChart {
  display: flex;
  justify-content: center;
  align-content: flex-start;
  flex-flow: wrap;  
}
