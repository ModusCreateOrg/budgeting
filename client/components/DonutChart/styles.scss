.donutChart {
  display: flex;
  height: 100%;  
  justify-content: center;
  align-content: flex-start;
  flex-flow: wrap;  
}
