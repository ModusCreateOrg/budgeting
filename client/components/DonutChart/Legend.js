import React, { PropTypes } from 'react';

import styles from './styles.scss';

const Legend = ({ data, color, dataLabel, dataKey }) => (
  <ul className={styles.legend}>
    {data.map(
      (item, idx) => <li style={{ color: color(idx) }} key={item[dataKey]}>{item[dataLabel]}</li>
    )}
  </ul>
);

Legend.propTypes = {
  data: PropTypes.array.isRequired,
  dataLabel: PropTypes.string.isRequired,
  dataKey: PropTypes.string.isRequired,
  color: PropTypes.func.isRequired,
};

export default Legend;
