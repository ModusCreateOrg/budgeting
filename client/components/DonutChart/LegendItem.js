import React, { PropTypes } from 'react';

import formatAmount from 'utils/formatAmount';
import styles from './styles.scss';

const LegendItem = ({ color, label, value }) => (
  <li style={{ color }}>
    <span>{label}</span>
    <span className={styles.value}> {formatAmount(value).text} </span>
  </li>
);

LegendItem.propTypes = {
  color: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
};

export default LegendItem;
