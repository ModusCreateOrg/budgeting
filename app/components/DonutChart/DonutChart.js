// @flow

import * as React from 'react';
import Legend from 'components/Legend';
import Chart from 'components/Chart';
import type { TransactionSummary } from 'selectors/transactions';
import { arc, pie, scaleOrdinal, schemeCategory10 } from 'd3';
import { shuffle } from 'utils/array';
import Path from './Path';
import styles from './styles.scss';

const randomScheme = shuffle(schemeCategory10);

type DonutChartProps = {
  data: TransactionSummary[],
  dataLabel: string,
  dataKey: string,
  dataValue?: string,
  color?: Function,
  height?: number,
  innerRatio?: number,
  useFormat?: boolean,
};

class DonutChart extends React.Component<DonutChartProps> {
  static defaultProps = {
    color: scaleOrdinal(randomScheme),
    height: 300,
    innerRatio: 4,
    dataValue: 'value',
    useFormat: true,
  };

  componentWillMount() {
    this.updateChartVariables();
  }

  componentWillReceiveProps(nextProps: DonutChartProps) {
    const { data, color, height } = nextProps;

    const old = this.props;

    if (old.data !== data || old.color !== color || old.height !== height) {
      this.updateChartVariables();
    }
  }

  getPathArc = () => {
    const { height, innerRatio } = this.props;
    return arc()
      .innerRadius(Number(height) / Number(innerRatio))
      .outerRadius(Number(height) / 2);
  };

  chart: any;

  pathArc: any;

  colorFn: any;

  outerRadius: number;

  boxLength: number;

  chartPadding = 8;

  updateChartVariables = () => {
    const { data, dataValue, color, height } = this.props;

    this.chart = pie()
      .value(d => d[dataValue])
      .sort(null);
    this.outerRadius = Number(height) / 2;
    this.pathArc = this.getPathArc();
    this.colorFn = color && color.domain && color.domain([0, data.length]);
    this.boxLength = height + this.chartPadding * 2;
  };

  render() {
    const { data, dataLabel, dataValue, dataKey, useFormat } = this.props;
    const { outerRadius, pathArc, colorFn, boxLength, chartPadding } = this;

    return (
      <div className={styles.donutChart}>
        <Chart
          width={boxLength}
          height={boxLength}
          padding={chartPadding}
          transform={`translate(${outerRadius},${outerRadius})`}
        >
          {this.chart(data).map((datum, index) => (
            <Path data={datum} index={index} fill={colorFn(index)} arcFn={pathArc} key={datum.data[dataKey]} />
          ))}
        </Chart>

        <Legend color={colorFn} {...{ data, dataValue, dataLabel, dataKey, useFormat }} />
      </div>
    );
  }
}

export default DonutChart;
