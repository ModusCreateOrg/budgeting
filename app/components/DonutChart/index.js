// @flow

import * as React from 'react';
import Legend from 'components/Legend';
import Chart from 'components/Chart';
import type { TransactionSummary } from 'selectors/transactions';
import { arc, pie, scaleOrdinal, schemeCategory20 } from 'd3';
import { shuffle } from 'utils/array';
import Path from './Path';
import styles from './styles.scss';

const randomScheme = shuffle(schemeCategory20);

type DonutChartProps = {
  data: TransactionSummary[],
  dataLabel: string,
  dataKey: string,
  dataValue: string,
  color: Function,
  height: number,
  innerRatio: number,
};

class DonutChart extends React.Component<DonutChartProps> {
  static defaultProps = {
    color: scaleOrdinal(randomScheme),
    height: 300,
    innerRatio: 4,
    dataValue: 'value',
  };

  componentWillMount() {
    this.updateChartVariables();
  }

  componentWillReceiveProps(nextProps: DonutChartProps) {
    const { data, color, height } = nextProps;

    const old = this.props;

    if (old.data !== data || old.color !== color || old.height !== height) {
      this.updateChartVariables();
    }
  }

  getPathArc = () => {
    const { height, innerRatio } = this.props;
    return arc()
      .innerRadius(height / innerRatio)
      .outerRadius(height / 2);
  };

  chart: any;
  pathArc: any;
  colorFn: any;
  outerRadius: number;
  boxLength: number;
  chartPadding = 8;

  updateChartVariables = () => {
    const { data, dataValue, color, height } = this.props;

    this.chart = pie()
      .value(d => d[dataValue])
      .sort(null);
    this.outerRadius = height / 2;
    this.pathArc = this.getPathArc();
    this.colorFn = color.domain && color.domain([0, data.length]);
    this.boxLength = height + this.chartPadding * 2;
  };

  render() {
    const { data, dataLabel, dataValue, dataKey } = this.props;
    const { outerRadius, pathArc, colorFn, boxLength, chartPadding } = this;

    return (
      <div className={styles.donutChart}>
        <Chart
          width={boxLength}
          height={boxLength}
          padding={chartPadding}
          transform={`translate(${outerRadius},${outerRadius})`}
        >
          {this.chart(data).map((datum, index) => (
            <Path data={datum} index={index} fill={colorFn(index)} arcFn={pathArc} key={datum.data[dataKey]} />
          ))}
        </Chart>

        <Legend color={colorFn} {...{ data, dataValue, dataLabel, dataKey }} />
      </div>
    );
  }
}

export default DonutChart;
