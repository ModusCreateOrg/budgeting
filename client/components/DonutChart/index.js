import React, { Component, PropTypes } from 'react';

import {
  arc,
  pie,
  scaleSequential,
  interpolateMagma
} from 'd3';

import Path from './Path';
import Legend from './Legend';
import styles from './styles.scss';


class DonutChart extends Component {
  static propTypes = {
    data: PropTypes.array.isRequired,
    dataValue: PropTypes.string.isRequired,
    dataLabel: PropTypes.string.isRequired,
    dataKey: PropTypes.string.isRequired,
    color: PropTypes.func,
    height: PropTypes.number,
    innerRatio: PropTypes.number,
  };

  static defaultProps = {
    color: scaleSequential().interpolator(interpolateMagma),
    height: 300,
    innerRatio: 4,
  };

  componentWillMount() {
    this.updateChartVariables();
  }

  componentWillReceiveProps({ data, color, height }) {
    const old = this.props;

    if (old.data !== data || old.color !== color || old.height !== height) {
      this.updateChartVariables();
    }
  }

  getPathArc = () => {
    const { height, innerRatio } = this.props;
    return arc().innerRadius(height / innerRatio).outerRadius(height / 2);
  }

  chartPadding = 8;

  updateChartVariables = () => {
    const { data, dataValue, color, height } = this.props;

    this.chart = pie().value(d => d[dataValue]).sort(null);
    this.outerRadius = height / 2;
    this.pathArc = this.getPathArc();
    this.colorFn = color.domain && color.domain([0, data.length]);
    this.boxLength = height + (this.chartPadding * 2);
    this.viewBox = `-${this.chartPadding} -${this.chartPadding} ${this.boxLength} ${this.boxLength}`;
  }

  render() {
    const { data, dataLabel, dataValue, dataKey } = this.props;
    const { outerRadius, pathArc, colorFn, boxLength, viewBox } = this;

    return (
      <div className={styles.donutChart}>
        <svg
          className={styles.mainSvg}
          width={boxLength}
          height={boxLength}
          viewBox={viewBox}
        >
          <g transform={`translate(${outerRadius},${outerRadius})`}>
            {this.chart(data).map(
              (datum, index) => (
                <Path
                  data={datum}
                  index={index}
                  fill={colorFn(index)}
                  arcFn={pathArc}
                  key={datum.data[dataKey]}
                />
              )
            )}
          </g>
        </svg>

        <Legend color={colorFn} {...{ data, dataValue, dataLabel, dataKey }} />
      </div>
    );
  }

}

export default DonutChart;
