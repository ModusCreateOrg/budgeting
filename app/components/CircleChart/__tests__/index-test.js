import React from 'react';
import renderer from 'react-test-renderer';
import DonutChart from '../';

// mock nested components
jest.mock('components/DonutChart/Path');
jest.mock('components/Chart', () => 'div');
jest.mock('components/Legend', () => 'div');

it('renders correctly', () => {
  const tree = renderer.create(<DonutChart dataLabel="test" dataKey="test" data={[]} />).toJSON();
  expect(tree).toMatchSnapshot();
});
