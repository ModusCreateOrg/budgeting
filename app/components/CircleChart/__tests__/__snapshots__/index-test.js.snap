// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="donutChart"
>
  <div
    height={316}
    padding={8}
    transform="translate(150,150)"
    width={316}
  />
  <div
    color={[Function]}
    data={Array []}
    dataKey="test"
    dataLabel="test"
    dataValue="value"
  />
</div>
`;
